### ISSUE TITLE
Enhancement: Semaphore/Lock/Mutex on "actuator/health" endpoint

### ISSUE THREAD

**Issue by rfelgent** [spring-projects/spring-boot#25459]
Hello,

it would be great that spring boot actuator provides a locking/semaphore/mutex logic out-of-the box to prevent its endpoints from flooding.
I have especially the "actuator/health" in mind, as this endpoint is extendable by custom HealthIndicators with potential problems. One of my HealthIndicators caused a _long_ running task by invocing another service/repository (in the end: a missing response/connection timeout fixed the trouble to some degree). As the calculation took so long, my monitoring system (consul) fired more requests in order to get a state. But that only flooded my system...

Currently, there is a caching implemented for th actuator endpoints: once the value is calculated it will be cached. The missing piece for me is the semaphor/lock mechanism for not yet calculated endpoints. This would help to avoid a flooding of the system.

I was wondering if such a feature could be provided by SpringBoot Actuator (like the caching...) and wanted to reach you out before implementing a custom solution on my own...

Best regards.

**Comment by wilkinsona** [comment]
Something like this would help with ActiveMQ's failover support where checking the health of a connection can block indefinitely. See https://github.com/spring-projects/spring-boot/issues/25188#issuecomment-782231036 for some details.

**Comment by rfelgent** [comment]
Hi @wilkinsona,

I will start to implement something and will get back to you as soon as the first draft is ready

**Comment by wilkinsona** [comment]
@rfelgent Just to set expectations, my comment was primarily intended just to link the two issues together. It doesn't necessarily mean that we'll decide to add this sort of functionality to Spring Boot. As I said in my comment on #25188, it'll add quite a bit of complexity and we'll need to weigh up the pros and cons. There are also a few different approaches that could be taken. If you'd like to implement something and share it with us, that would be great. I just want you to be aware that we may decide not to do anything in this area or decide to take a different approach to whatever you may propose.

**Comment by rfelgent** [comment]
Hi @wilkinsona ,

thx for clearifying "my expectation".  :-)

My main goal was to ensure that I did not miss this feature in spring actuator itself.
It seems that this feature is not avialable at the moment and that it _might_ be valuable for others (thx for the link to issue #25188) 

Anyway, I prefer a lock/semaphore feature provided by Spring Boot Actuator, as it would save me from managing custom code in my app or a custom dependency.

I would like to talk about the functionality in general and my ideas of an implementation with somebody from Spring Boot team, so that I can push that topic. 
I am totally fine if either the functionality (or its implementation) is rejected - but we should talk about it first, shouldn't we ?

Best regards

**Comment by wilkinsona** [comment]
From #2652:

>  I've been wondering about something that runs the health indicators periodically in the background and the health endpoint would then just respond immediately with the current aggregated status.

**Comment by rfelgent** [comment]
Hi @wilkinsona ,

sorry for my abscence last days!

I have noticed, that you have changed the topic. The my original idea of a "mutex/lock" functionality is dropped in favour of an asynchronous, periodically running task ?
 
Can I conclude, that this is the favourite design/concept ?

**Comment by wilkinsona** [comment]
I was imagining that a request to the health endpoint would immediately return the current state and something else would keep this current state up-to-date. This could be done by scheduling something that runs periodically in the background, or on demand when a request is received. In the on-demand case, we'd probably need a configurable maximum period that a request will wait for the health status to be updated before it responds with down or unknown.

**Comment by rfelgent** [comment]
does the concept include a dedicated scheduling for each healthcontributor with specific rates and/or delays?

**Comment by wilkinsona** [comment]
I'm afraid we don't know yet, hence the pending design work label. My initial reaction is that we'd want to avoid that level of complexity and, if we took a scheduling-based approach, we'd want to use one schedule for all of the contributors.

**Comment by 9-9-0** [comment]
Is the design work still pending?

**Comment by wilkinsona** [comment]
Yes, it is. We may take a look at it in the next month or two as part of planning Spring Boot 2.6.

**Comment by denisa** [comment]
Could https://github.com/antoinemeyer/spring-boot-async-health-indicator serves as inspiration?

**Comment by wilkinsona** [comment]
Thanks, @denisa. There has been some discussion of that on https://github.com/spring-projects/spring-boot/issues/2652.
